### PR TITLE
[Test, BugFix] add regression test for PR #203

### PR DESCRIPTION
**Pull request by apbard**
#203 fixed a bug but didn't include a dedicated regression test

### CODE REVIEW

File: test/test_tensorclass_nofuture.py [hunk 1]
```diff
@@ -354,6 +354,47 @@ class MyDataNested:
...
+        batch_size=[3, 4],
+    )
+    data2_wrong = MyDataMemMap2(
+        x=MemmapTensor.from_tensor(torch.ones(3, 4, 5)),
+        y=torch.ones(3, 4, 5),
+        batch_size=[3, 4],
+    )
+    with pytest.raises(ValueError, match="__setitem__ is only allowed for same-class assignement"):
+        data1[:2] = data2_wrong[:2]
+    with pytest.raises(ValueError, match="__setitem__ is only allowed for same-class assignement"):
```
Comment (apbard) [inline test/test_tensorclass_nofuture.py:381]: The fact that this raises and the following case not is a bit strange. @vmoens wdyt?

File: test/test_tensorclass_nofuture.py [hunk 2]
```diff
@@ -354,6 +354,47 @@ class MyDataNested:
...
+        data2_wrong[2:] = data1[2:]
+        
+    data2 = MyDataMemMap1(
+        x=MemmapTensor.from_tensor(torch.ones(3, 4, 5)),
+        y=torch.ones(3, 4, 5),
+        batch_size=[3, 4],
+    )    
+    data1[:2] = data2[:2]
+    assert (data1[:2] == 1).all()
```
Comment (apbard) [inline test/test_tensorclass_nofuture.py:390]: this currently returns False. I think we should treat scalars differently and perform a broadcast @vmoens, @tcbegley, @sreevasthav wdyt?

Comment (vmoens) [reply test/test_tensorclass_nofuture.py:390]: Yep
Happy to implement some broadcasting for `TensorDIct` first with float, integers, tensors and memmap-tensors in `__setitem__`.

File: test/test_tensorclass_nofuture.py (hunk 1 above)
Comment (vmoens) [reply test/test_tensorclass_nofuture.py:381]: I just tried and on my side both tests pass

File: test/test_tensorclass_nofuture.py (hunk 2 above)
Comment (vmoens) [reply test/test_tensorclass_nofuture.py:390]: This actually works as of today
```python
>>> from tensordict import TensorDict
td = TensorDict>>> td = TensorDict({"a": [1, 1, 1]}, [])
>>> td==1
TensorDict(
    fields={
        a: Tensor(shape=torch.Size([3]), device=cpu, dtype=torch.bool, is_shared=False)},
    batch_size=torch.Size([]),
    device=None,
    is_shared=False)
>>> (td==1)["a"]
tensor([True, True, True])
```
but this does not 
```
>>> td==torch.zeros(3)
False
```
I can make a quick patch

Comment (vmoens) [reply test/test_tensorclass_nofuture.py:390]: I also think it makes sense to compare two tensorclass of a different type, but I might get this wrong.
Our type checks work with `isinstance`, hence we'll get some weird behaviours if one class inherits from another on the lhs or rhs (ie `a==b` is not `b==a`)

File: test/test_tensorclass_nofuture.py (hunk 1 above)
Comment (apbard) [reply test/test_tensorclass_nofuture.py:381]: they pass *without* the pytest.raise?

File: test/test_tensorclass_nofuture.py (hunk 2 above)
Comment (apbard) [reply test/test_tensorclass_nofuture.py:390]: > I also think it makes sense to compare two tensorclass of a different type, but I might get this wrong. Our type checks work with `isinstance`, hence we'll get some weird behaviours if one class inherits from another on the lhs or rhs (ie `a==b` is not `b==a`)

we can relax the constraint to be just a tensorclass with all keys equal

File: test/test_tensorclass_nofuture.py (hunk 1 above)
Comment (vmoens) [reply test/test_tensorclass_nofuture.py:381]: I'm confused, the first should work with `pytest.raises` but not the second, which should not raise any exception. Is it what you're saying? Why would it be the case?

Comment (apbard) [reply test/test_tensorclass_nofuture.py:381]: no, i am comparing this two "raises" checks with the case at lines 389 and following. The only difference between data2_wrong and data2 is the class name. I think that is similar to what you mention about inheritance in the other comment thread.

File: test/test_tensorclass_nofuture.py [hunk 3]
```diff
@@ -354,6 +354,52 @@ class MyDataNested:
...
+    )
+    data2_wrong = MyDataMemMap2(
+        x=MemmapTensor.from_tensor(torch.ones(3, 4, 5)),
+        y=torch.ones(3, 4, 5),
+        batch_size=[3, 4],
+    )
+    with pytest.raises(
+        ValueError, match="__setitem__ is only allowed for same-class assignement"
+    ):
+        data1[:2] = data2_wrong[:2]
```
Comment (vmoens) [inline test/test_tensorclass_nofuture.py:382]: it's not super clear to me why we don't want to allow this?
We should be able to write a tensor on a memmap-tensor and inversely

Comment (apbard) [reply test/test_tensorclass_nofuture.py:382]: Here we are actually writing a `MyDataMemMap2` on `MyDataMemMap1` and it is currently forbidden (hence the pytest.raises), but I am totally in favour to change this behaviour as per previous discussion.

Comment (vmoens) [reply test/test_tensorclass_nofuture.py:382]: Yeah I would expect it to work, or at least not be disabled by design

Comment (vmoens) [reply test/test_tensorclass_nofuture.py:382]: Should I do it? Do we want to merge this first?

Comment (apbard) [reply test/test_tensorclass_nofuture.py:382]: I would merge. I can have a look at that

**Review by vmoens — APPROVED** [review]

**Pull request merged**